### PR TITLE
Vivax set equilibrium

### PR DESCRIPTION
**Pull request by RJSheppard**
Vivax solution, complementary with malariaEquilibriumVivax.

### CODE REVIEW

File: R/human_infection.R [hunk 1]
```diff
@@ -209,37 +209,57 @@ calculate_infections <- function(
...
-    rate_new_infections_vector[with_hypnozoites_vector] <- rate_new_infections_vector[with_hypnozoites_vector] + rate_relapse
-    rate_new_infections_vector <- rate_new_infections_vector[rate_new_infections_vector>0]
-    # Convert total rate of new infection to probability
-    prob_new_infections_vector <- rate_to_prob(rate_new_infections_vector)
+    prob_new_infections <- rate_to_prob(rate_infection_complete)
 
     # Subset for new infections/bite infections
-    newly_infected <- bitset_at(potential_new_infections_humans, bernoulli_multi_p(prob_new_infections_vector))
+    newly_infected <- bitset_at(potential_new_infections_humans,
+                                bernoulli_multi_p(prob_new_infections[potential_new_infections_humans$to_vector()]))
```
Comment (giovannic) [inline R/human_infection.R:232]:
```suggestion
      bernoulli_multi_p(prob_new_infections[potential_new_infections_humans$to_vector()]))
```

File: R/human_infection.R [hunk 2]
```diff
@@ -209,37 +209,57 @@ calculate_infections <- function(
...
-    # Add new batches for new bite infections
-    variables$hypnozoites$queue_update(
-      variables$hypnozoites$get_values(newly_bite_infected) + 1,
-      newly_bite_infected
-    )
+    ## All bitten humans with an infectious bite (incorporating prophylaxis) get a new batch of hypnozoites
+    if(newly_bite_infected$size()>0){
+      new_batches_formed <- bitset_at(newly_bite_infected, bernoulli_multi_p(1-ls_prophylaxis))
+      variables$hypnozoites$queue_update(
+        # variables$hypnozoites$get_values(new_batches_formed$and(can_increase_batches)) + 1,
```
Comment (giovannic) [inline R/human_infection.R:258]:
```suggestion
```

File: R/variables.R [hunk 3]
```diff
@@ -185,7 +193,8 @@ create_variables <- function(parameters) {
       )
     )
 
-    # Hypnozoite batch variable
+    # Hypnozoite - initial immunity function used to initiate hypnozoites
+    # browser()
```
Comment (giovannic) [inline R/variables.R:197]:
```suggestion
```

File: R/variables.R [hunk 4]
```diff
@@ -352,6 +361,7 @@ initial_immunity <- function(
     return(vnapply(
       seq_along(age),
       function(i) {
+        # browser()
```
Comment (giovannic) [inline R/variables.R:364]:
```suggestion
```

File: R/variables.R [hunk 5]
```diff
@@ -415,17 +427,27 @@ calculate_initial_counts <- function(parameters) {
...
+          parameters$init_EIR * calculate_zeta(n, parameters),
+          ft,
+          parameters$eq_params,
+          EQUILIBRIUM_AGES
+        )
+      }
+    )
+
+  } else if (parameters$parasite == "vivax"){
+    eq <- malariaEquilibriumVivax::vivax_equilibrium(
```
Comment (giovannic) [inline R/variables.R:444]: Can you add malariaEquilibriumVivax to DESCRIPTION? In the same way malariaEquilibrium is

File: tests/testthat/test-biology-vivax.R [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    parameters$mum
+  )
+
+  mockery::expect_args(
+    events$mosquito_death$schedule,
+    1,
+    dead,
+    0
+  )
+})
```
Comment (giovannic) [inline tests/testthat/test-biology-vivax.R:250]: These tests don't depend on falciparum or vivax right? We can just remove them

File: R/human_infection.R [hunk 7]
```diff
@@ -209,37 +209,57 @@ calculate_infections <- function(
...
 
     # Subset for new infections/bite infections
-    newly_infected <- bitset_at(potential_new_infections_humans, bernoulli_multi_p(prob_new_infections_vector))
+    newly_infected <- bitset_at(potential_new_infections_humans,
+                                bernoulli_multi_p(prob_new_infections[potential_new_infections_humans$to_vector()]))
     newly_bite_infected <- bitset_at(
-      potential_new_infections_humans,
-      bernoulli_multi_p(rate_bitten/rate_new_infections_vector[source_vector]))$and(newly_infected)
+      newly_infected,
+      bernoulli_multi_p(relative_rate[newly_infected$to_vector()]))
```
Comment (giovannic) [inline R/human_infection.R:235]: This looks like it's going to slow vivax computations down a lot :(. There are a lot of very slow `$to_vector` operations.

But it's unclear to me how to speed this up on first look, but I'll try again tomorrow.

File: R/human_infection.R (hunk 1 above)
Comment (RJSheppard) [reply R/human_infection.R:232]: I've adjusted the indentation slightly differently in the next commit.

File: R/variables.R (hunk 5 above)
Comment (RJSheppard) [reply R/variables.R:441]: Added in the next commit.

File: tests/testthat/test-biology-vivax.R (hunk 6 above)
Comment (RJSheppard) [reply tests/testthat/test-biology-vivax.R:250]: Good shout! Removed in the next commit.

File: R/human_infection.R [hunk 8]
```diff
@@ -207,42 +209,32 @@ calculate_infections <- function(
...
     newly_bite_infected <- bitset_at(
-      newly_infected,
-      bernoulli_multi_p(relative_rate[newly_infected$to_vector()]))
+      source_humans,
+      bernoulli_multi_p(relative_rate))$and(newly_infected)
 
     ## Drug prophylaxis may limit formation of new hypnozoite batches
-    newly_bite_infected_vector <- newly_bite_infected$to_vector()
-    ls_prophylaxis <- rep(0, length(newly_bite_infected_vector))
+    # newly_bite_infected_vector <- newly_bite_infected$to_vector()
```
Comment (giovannic) [inline R/human_infection.R:230]:
```suggestion
```

**Review by giovannic — CHANGES_REQUESTED** [review]
This looks much better!

After these requested changes are made, I would double check that your new biology tests pass locally before merging this in.

**Pull request merged**